### PR TITLE
[14.0][ADD] mrp_subcontracting_purchase

### PR DESCRIPTION
**Pull request by Volodiay622**

### CODE REVIEW

**Comment by francesco-ooops** [comment]
@pedrobaeza could you please provide us instructions on how to correctly bring to merge this PR? 

eg. regarding authorship and commits, as this is the backporting to v14 of odoo module https://github.com/odoo/odoo/tree/15.0/addons/mrp_subcontracting_purchase

Thanks!

**Comment by pedrobaeza** [comment]
First, you should add a commit attributed to "Odoo SA <odoo@odoo.com>" with the contents as they are in the new version, and then add your commits with the changes needed to work on this version.

Then, check https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#10backporting-odoo-modules

**Comment by Aldeigja** [comment]
> First, you should add a commit attributed to "Odoo SA [odoo@odoo.com](mailto:odoo@odoo.com)" with the contents as they are in the new version, and then add your commits with the changes needed to work on this version.
> 
> Then, check https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#10backporting-odoo-modules

@pedrobaeza To make mrp_subcontracting_purchase work as per v15, we also had to port code from v15 of Odoo modules:

mrp_subcontracting_purchase, 
mrp_subcontracting_dropshipping,
purchase_mrp,
purchase_stock

how do you suggest to proceed for code from those modules?

**Comment by pedrobaeza** [comment]
You have to include the needed changes in your module, trying to be the most compatible possible. I needed to do similar things when I backported `mrp_subcontracting`. Some examples:

https://github.com/OCA/manufacture/commit/f32ce8ba46c65ab3385ecaef82bd05fcb4509f41#diff-96c2f57c6a07eb25b5e8b5cbc49574423aa16aa8fbb1c50672de67d9a25aca56R150
https://github.com/OCA/manufacture/commit/f32ce8ba46c65ab3385ecaef82bd05fcb4509f41#diff-96c2f57c6a07eb25b5e8b5cbc49574423aa16aa8fbb1c50672de67d9a25aca56L204
https://github.com/OCA/manufacture/commit/f32ce8ba46c65ab3385ecaef82bd05fcb4509f41#diff-024c81702831beef5a3aaf69d8fb0342d470277883158ee96b248722f102a33bR20

File: mrp_subcontracting_purchase/models/stock_picking.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        moves_subcontracted = raw_mp_id.move_finished_ids.move_dest_ids.filtered(
+            lambda m: m.is_subcontract
+        )
+        return moves_subcontracted.purchase_line_id.order_id
+
+    def _get_subcontract_production(self):
+        """Returns subcontract production in stock picking line"""
+        return self.move_lines._get_subcontract_production()
+
+    def _action_done(self):
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_picking.py:60]: This function is too complex. Please function code split on functions.

Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_picking.py:60]: This function is too complex. Please function code split on functions.

File: mrp_subcontracting_purchase/models/stock_rule.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            )
+            procurements_by_po_domain[domain].append((procurement, rule))
+        if errors:
+            raise ProcurementException(errors)
+        self._create_po_not_exist(procurements_by_po_domain)
+
+    def _prepare_purchase_order(self, company_id, origins, values):
+        """Returns prepared data for create PO"""
+        if (
+            "partner_id" not in values[0]
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_rule.py:49]: `value` may be empty? If `value` will be empty than you need check this before this condition.

File: mrp_subcontracting_purchase/models/stock_rule.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                po = (
+                    po_obj.with_company(company_id).with_user(SUPERUSER_ID).create(vals)
+                )
+
+        else:
+            # If a purchase order is found, adapt its `origin` field.
+            if po.origin:
+                missing_origins = origins - set(po.origin.split(", "))
+                if missing_origins:
+                    po.write({"origin": po.origin + ", " + ", ".join(missing_origins)})
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_rule.py]: Could you use `format` function for preparing string value?

File: mrp_subcontracting_purchase/models/mrp_production.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            .filtered(lambda m: m.state not in ("done", "cancel"))
+            ._action_record_components()
+        )
+        action["res_id"] = backorder.id
+        return action
+
+    def _pre_button_mark_done(self):
+        if self._get_subcontract_move():
+            return True
+        return super()._pre_button_mark_done()
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/mrp_production.py]: You can do like this (optional):
```
return True if self._get_subcontract_move() else super()._pre_button_mark_done()
```

File: mrp_subcontracting_purchase/models/mrp_production.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    def _subcontracting_filter_to_done(self):
+        # OVERRIDE, to add condition 'not mo.subcontracting_has_been_recorded',
+        # which checks whether the subcontracting has been recorded or not
+
+        def filter_in(mo):
+            if (
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/mrp_production.py]: You can inverse this big condition and write `return` without if:
```
return  not (mo.state in ("done", "cancel") .......)`

File: mrp_subcontracting_purchase/models/purchase_order.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            elif not order_line.display_type:
+                not_in_bom_products |= order_line
+        if in_bom_products and not_in_bom_products:
+            raise UserError(
+                _(
+                    """It appears some components in this RFQ are not meant for subcontracting.
+                    Please create a separate order for these."""
+                )
+                + "\n\n"
+                + "\n".join(not_in_bom_products.mapped("name"))
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/purchase_order.py:119]: It's not good idea, you must use format like this for one variable `_("....... \n\n%s", value)`.

File: mrp_subcontracting_purchase/models/stock_move.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            if (
+                productions._has_tracked_component()
+                or productions[:1].consumption != "strict"
+            ):
+                move.display_assign_serial = False
+
+    def _compute_show_subcontracting_details_visible(self):
+        """Compute if the action button in order to see moves raw is visible"""
+        self.show_subcontracting_details_visible = False
+        for move in self:
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_move.py:30]: I think need add the filter for `self`
```py
for move in self.filtered("is_subcontract"):
    ....
```

File: mrp_subcontracting_purchase/models/stock_move.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            ):
+                continue
+            move.show_subcontracting_details_visible = True
+
+    def _compute_show_details_visible(self):
+        """If the move is subcontract and the components are tracked. Then the
+        show details button is visible.
+        """
+        res = super(StockMove, self)._compute_show_details_visible()
+        for move in self:
```
Comment (geomer198) [inline mrp_subcontracting_purchase/models/stock_move.py:48]: Here too

**Review by geomer198 — COMMENTED** [review]
Please check my comments and upper your tests coverage.

File: mrp_subcontracting_purchase/models/stock_rule.py (hunk 2 above)
Comment (Volodiay622) [reply mrp_subcontracting_purchase/models/stock_rule.py:49]: value is checking [here](https://github.com/OCA/manufacture/pull/1082/files/adf1f14f9e0686498b279baa8cd7fd1b899aa259#diff-b8f74f699c3783e44e9e4347460bac8b9c835865069caa66f1fdce52d8f9385bR182-R188)

File: mrp_subcontracting_purchase/models/mrp_production.py (hunk 4 above)
Comment (Volodiay622) [reply mrp_subcontracting_purchase/models/mrp_production.py]: applied, thanks

File: mrp_subcontracting_purchase/models/stock_rule.py (hunk 3 above)
Comment (Volodiay622) [reply mrp_subcontracting_purchase/models/stock_rule.py]: resolved

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

We should be good here, full test coverage and documentation

@pedrobaeza can we ask for your feedback? Thanks!

**Comment by francesco-ooops** [comment]
@david-fleischmann would this help your flow?

**Comment by pedrobaeza** [comment]
I'm not seeing an initial commit authored by Odoo SA with the content as is on the next version, and then, all your modifications to make it work in v14.

**Comment by geomer198** [comment]
@pedrobaeza Could you check repository structure please?
Now is correct?

File: mrp_subcontracting_purchase/README.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+This module is a backport from Odoo SA and as such, it is not included in the OCA CLA.
```
Comment (pedrobaeza) [inline mrp_subcontracting_purchase/README.rst]:
```suggestion
This bridge module adds some smart buttons between Purchase and Subcontracting

**DISCLAIMER:** This module is a backport from Odoo SA and as such, it is not included in the OCA CLA.
```

File: mrp_subcontracting_purchase/demo/0001_demo_product_category.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<odoo>
```
Comment (pedrobaeza) [inline mrp_subcontracting_purchase/demo/0001_demo_product_category.xml]: Why do you put a number on the file names?

Comment (pedrobaeza) [reply mrp_subcontracting_purchase/demo/0001_demo_product_category.xml]: The proper name is `product_category_demo.xml`.

File: mrp_subcontracting_purchase/README.rst (hunk 9 above)
Comment (pedrobaeza) [reply mrp_subcontracting_purchase/README.rst]: Oh, I see that you have a DESCRIPTION.rst in the `readme` folder, so you have to introduce the disclaimer there.

Comment (geomer198) [reply mrp_subcontracting_purchase/README.rst]: > Oh, I see that you have a DESCRIPTION.rst in the `readme` folder, so you have to introduce the disclaimer there.

Resolved

File: mrp_subcontracting_purchase/demo/0001_demo_product_category.xml (hunk 10 above)
Comment (geomer198) [reply mrp_subcontracting_purchase/demo/0001_demo_product_category.xml]: > The proper name is `product_category_demo.xml`.

Files name are changed.

File: mrp_subcontracting_purchase/README.rst [hunk 11]
```diff
@@ -0,0 +1,5 @@
+This bridge module adds some smart buttons between Purchase and Subcontracting
+
+**DISCLAIMER:** This module is a backport from Odoo SA and as such, it is not included in the OCA CLA.
```
Comment (pedrobaeza) [inline mrp_subcontracting_purchase/README.rst]: This should be put on `readme/DESCRIPTION.rst`, not here.

**Comment by francesco-ooops** [comment]
Hi @pedrobaeza , thanks for your support on this one. 

Do you have any idea why the "Configuration" file is not appearing in readme here?: https://github.com/OCA/manufacture/tree/14.0/mrp_subcontracting_purchase

Do you think it's a .md issue?

**Comment by pedrobaeza** [comment]
It should be `CONFIGURE.md`, not `CONFIGURATION.md`.

**Comment by david-fleischmann** [comment]
@francesco-ooops thank you! Just found that notifications from github in my spam folder... sorry for that late reply!
I'll check this feature on our side within the next weeks. Best, David

**Pull request merged**